src/utils/marketsAndNetworksConfig.ts

```diff
@@ -143,6 +145,134 @@ export const isFeatureEnabled = {
+
+        this.emit('network', network, null);
+      }
+    }
+    return network;
+  }
+} */
+
+/**
+ * The provider will rotate rpcs on error.
+ * If provider rotates away from the first RPC, rotate back after a set interval to prioritize using most reliable RPC.
+ * If provider rotates through all rpcs, delay to avoid spamming rpcs with requests.
+ */
+class RotationProvider extends ethersProviders.BaseProvider {
+  readonly providers: StaticJsonRpcProvider[];
+  private currentProviderIndex = 0;
+  private lastFullRotationTimestamp = 0;
+  private firstRotationTimestamp = 0;
+  // after completing a full rotation of the RotationProvider, delay to avoid spamming rpcs with requests
+  private ROTATION_DELAY = 5000;
+  // if we rotate away from first rpc, return back after this delay
+  private FALL_FORWARD_DELAY = 60000;
+
+  constructor(urls: string[], chainId: number) {
+    super(chainId);
+    this.providers = urls.map((url) => new StaticJsonRpcProvider(url, chainId));
+  }
+
+  /**
+   * If provider rotates through all rpcs, delay to avoid spamming rpcs with requests
+   */
+  async delayRotation() {
+    const now = new Date().getTime();
+    const diff = now - this.lastFullRotationTimestamp;
+    if (diff < this.ROTATION_DELAY) {
+      await sleep(this.ROTATION_DELAY - diff);
+    }
+  }
+
+  /**
+   * If we rotate away from the first RPC, rotate back after a set interval to prioritize using most reliable RPC
+   */
+  async fallForwardRotation() {
+    const now = new Date().getTime();
+    const diff = now - this.firstRotationTimestamp;
+    if (diff < this.FALL_FORWARD_DELAY) {
+      await sleep(this.FALL_FORWARD_DELAY - diff);
+      this.currentProviderIndex = 0;
+    }
+  }
+
+  /**
+   * If rpc fails, rotate to next available and trigger rotation or fall forward delay where applicable
+   * @param prevIndex last updated index, checked to avoid having multiple active rotations
+   */
+  private async rotateUrl(prevIndex: number) {
+    // don't rotate when another rotation was already triggered
+    if (prevIndex !== this.currentProviderIndex) return;
+    // if we rotate away from the first url, switch back after FALL_FORWARD_DELAY
+    if (this.currentProviderIndex === 0) {
+      this.currentProviderIndex += 1;
+      this.firstRotationTimestamp = new Date().getTime();
+      this.fallForwardRotation();
+    } else if (this.currentProviderIndex === this.providers.length - 1) {
+      await this.delayRotation(); // wait for ROTATION_DELAY after completing a full rotation
+      this.currentProviderIndex = 0;
+      this.lastFullRotationTimestamp = new Date().getTime();
+    } else {
+      this.currentProviderIndex += 1;
+    }
+  }
+
+  async detectNetwork(): Promise<Network> {
+    const index = this.currentProviderIndex;
+    try {
+      return await this.providers[index].detectNetwork();
+    } catch (e) {
+      console.log(e.message);
+      await this.rotateUrl(index);
+      return this.detectNetwork();
```

Comment:
I'm not sure that this function should be recursive. I think we should only be checking the network for the current provider, and either return or throw. If it throws, I think we'll handle it in the action other action methods and then do the rotate.

Reply:
I guess my assumption is is that the base class is calling this when we're trying a provider, and if we switch providers during the middle of that, we'll have multiple retry's going on

Reply:
I think we need to dig into where `detectNetwork` is being called to decide what to do here.

My understanding is that the `StaticJsonRpcProvider` does not call this method when performing an action, only on initialization so this could be dangerous if the error is uncaught and a `perform` or `send` is called on the wrong network

Reply:
I was looking at this a little more, it was a different issue that was causing the weird behavior on failing that's fixed now.

`detectNetwork` should only be called on initialization. And since we're using static json rpc providers, I think it'd make sense for this method to just call `detectNetwork` on all of the providers. Maybe we want to copy some of the logic from the `Fallback` provider on how it handles that.

I don't think it makes sense to call `rotateUrl` here. Should just be something like:
```
async detectNetwork() {
  return await this.providers.map(p => p.detectNetwork());
}
```
Might now exactly like that, but we just need to verify that all provider networks have the same configured chainId

Reply:
sorry somehow missed.
Do however you please, has been a while since i touched this and would need to have a deeper look.
Iirc when you initialize via `StaticJsonProvider(rpc, networkId)` `detectNetwork` should never fail as it just resolves the static networkId inside `this._networkId`.